src/main/java/com/hazelcast/aws/impl/AwsOperation.java

```diff
@@ -0,0 +1,332 @@
+
+    /**
+     * @param args
+     */
+    protected void prepareHttpRequest(Object... args) {
+        body = "";
+    }
+
+    /**
+     * Add available filters to narrow down the scope of the query
+     * Override where appropriate.
+     */
+    protected void addFilters() {
+    }
+
+    /**
+     * AWS response codes for client and server errors are specified here:
+     * {@see http://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html}.
+     */
+    private static boolean isAwsError(int responseCode) {
+        return responseCode >= MIN_HTTP_CODE_FOR_AWS_ERROR && responseCode < MAX_HTTP_CODE_FOR_AWS_ERROR;
+    }
+
+    private static String extractErrorMessage(HttpURLConnection httpConnection) {
+        InputStream errorStream = httpConnection.getErrorStream();
+        if (errorStream == null) {
+            return "";
+        }
+        return readFrom(errorStream);
+    }
+
+    private static String readFrom(InputStream stream) {
+        Scanner scanner = new Scanner(stream, UTF8_ENCODING).useDelimiter("\\A");
+        return scanner.hasNext() ? scanner.next() : "";
+    }
+
+    // Visible for testing
+    void retrieveCredentials()
+            throws IOException {
+        if (isEcsEndpoint()) {
+            retrieveContainerCredentials(getEnvironment());
+        }
+        else {
+            retrieveIamRole();
+            if (isNotEmpty(awsCredentials.getIamRole())) {
+                retrieveIamRoleCredentials();
+            } else {
+                // legacy ECS mode
+                retrieveContainerCredentials(getEnvironment());
+            }
+        }
+    }
+
+
+    private void retrieveIamRole() {
+        if (isEmpty(awsCredentials.getIamRole()) || "DEFAULT".equals(awsCredentials.getIamRole())) {
+            String defaultIAMRole = null;
+            try {
+                defaultIAMRole = getDefaultIamRole();
+            } catch (Throwable e) {
+                LOGGER.finest("Cannot get DEFAULT IAM role. CONTINUING. Exception was: " + e.getMessage());
+            }
+            awsCredentials.setIamRole(defaultIAMRole);
+        }
+    }
+
+    private String getDefaultIamRole()
+            throws IOException {
+        String uri = INSTANCE_METADATA_URI.concat(IAM_SECURITY_CREDENTIALS_URI);
+        return retrieveRoleFromURI(uri);
+    }
+
+    private void retrieveIamRoleCredentials() {
+        try {
+            String query = IAM_SECURITY_CREDENTIALS_URI.concat(awsCredentials.getIamRole());
+            String uri = INSTANCE_METADATA_URI.concat(query);
+            String json = retrieveRoleFromURI(uri);
+            parseAndStoreRoleCreds(json);
+        } catch (Exception io) {
+            throw new InvalidConfigurationException("Unable to retrieve credentials from IAM Role: " + awsCredentials.getIamRole(),
+                    io);
+        }
+    }
+
+    private void retrieveContainerCredentials(Environment env)
+            throws IOException {
```

Comment:
We need to refactor that part because now we already asks users to define which environment they are deploying via host header.. ec2 or ecs...

Reply:
OK, if we have `ecs` we directly call this one